cpmpy/tools/dimacs.py

```diff
@@ -71,42 +71,49 @@ def write_dimacs(model, fname=None):
+        Read a CPMpy model from a DIMACS formatted file striclty following the specification: https://web.archive.org/web/20190325181937/https://www.satcompetition.org/2009/format-benchmarks2009.html
+        Note: the p-line has to denote the correct number of variables and clauses
         :param: fname: the name of the DIMACS file
         :param: sep: optional, separator used in the DIMACS file, will try to infer if None
     """
 
     m = cp.Model()
 
     with open(fname, "r") as f:
-
-        lines = f.readlines()
-        for i, line in enumerate(lines): # DIMACS allows for comments, skip comment lines
-            if line.startswith("p cnf"):
-                break
+        clause = []
+        had_p_line = False
+        for line in f.readlines():
+            if line == "" or line.startswith("c"):
+                continue  # skip empty and comment lines
+            elif line.startswith("p"):
+                try:
+                    _,typ,nr_vars,nr_cls = line.strip().split(" ")
+                except ValueError:
+                    raise cp.exceptions.CPMpyException(f"Invalid DIMACS file p-header: {line}")
+                if typ != "cnf":
+                    raise ValueError("Expected `cnf` (i.e. DIMACS) as file format, but got {typ} which is not supported.")
+                nr_vars = int(nr_vars)
+                if nr_vars>0:
+                    bvs = cp.boolvar(shape=nr_vars)
+                nr_cls = int(nr_cls)
+                had_p_line = True
             else:
-                assert line.startswith("c"), f"Expected comment on line {i}, but got {line}"
-
-        cnf = "\n".join(lines[i+1:]) # part of file containing clauses
-
-        bvs = []
-        txt_clauses = re.split(r"\n* \n*0", cnf) # clauses end with ` 0` but can have arbitrary newlines
-
-        for txt_ints in txt_clauses:
-            if txt_ints is None or len(txt_ints.strip()) == 0:
-                continue # empty clause or weird format
-
-            clause = []
-            ints = [int(idx.strip()) for idx in txt_ints.split(" ") if len(idx.strip())]
-
-            for i in ints:
-                if abs(i) >= len(bvs):  # var does not exist yet, create
-                    bvs += [cp.boolvar() for _ in range(abs(i) - len(bvs))]
-                bv = bvs[abs(i) - 1]
-                clause.append(bv if i > 0 else ~bv)
-
-            m += cp.any(clause)
+                assert had_p_line, "Expected p-line before clauses"
+                for token in line.strip().split():
+                    i = int(token.strip())
+                    if i == 0:
+                        m+=cp.any(clause)
+                        clause = []
+                    else:
+                        try:
+                            bv = bvs[abs(i)-1]
+                        except IndexError:
+                            raise AssertionError("Expected <={nr_vars} variables (from p-line) but found literal {i} in clause {line}")
```

Comment:
Hmm ValueError would be better here. AssertErrors are not supposed to be thrown explicilty I feel